### PR TITLE
bpo-35459: Use PyDict_GetItemWithError() instead of PyDict_GetItem()

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue35459

### CODE REVIEW

File: Objects/dictobject.c [hunk 1]
```diff
@@ -2358,14 +2371,14 @@ PyDict_MergeFromSeq2(PyObject *d, PyObject *seq2, int override)
         value = PySequence_Fast_GET_ITEM(fast, 1);
         Py_INCREF(key);
         Py_INCREF(value);
-        if (override || PyDict_GetItem(d, key) == NULL) {
-            int status = PyDict_SetItem(d, key, value);
-            if (status < 0) {
+        if (override || PyDict_GetItemWithError(d, key) == NULL) {
+            if ((!override && PyErr_Occurred()) || PyDict_SetItem(d, key, value) < 0) {
```
Comment (ZackerySpytz) [inline Objects/dictobject.c]: New code shouldn't violate PEP 7.

File: Modules/_csv.c [hunk 2]
```diff
@@ -1434,8 +1434,12 @@ csv_register_dialect(PyObject *module, PyObject *args, PyObject *kwargs)
 static PyObject *
 csv_unregister_dialect(PyObject *module, PyObject *name_obj)
 {
-    if (PyDict_DelItem(_csvstate_global->dialects, name_obj) < 0)
-        return PyErr_Format(_csvstate_global->error_obj, "unknown dialect");
+    if (PyDict_DelItem(_csvstate_global->dialects, name_obj) < 0) {
```
Comment (ericsnowcurrently) [inline Modules/_csv.c:1437]: Presumably this is a "drive-by" fix. :)

File: Modules/_sre.c [hunk 3]
```diff
@@ -1899,12 +1899,7 @@ match_getslice_by_index(MatchObject* self, Py_ssize_t index, PyObject* def)
     void* ptr;
     Py_ssize_t i, j;
 
-    if (index < 0 || index >= self->groups) {
-        /* raise IndexError if we were given a bad group number */
-        PyErr_SetString(
-            PyExc_IndexError,
-            "no such group"
-            );
+    if (index < 0) {
```
Comment (ericsnowcurrently) [inline Modules/_sre.c]: Was this change intentional?  At first glance it looks like the code you've removed actually matters.

Is this function only ever called with a pre-validated index (e.g. the one returned from `match_getindex()`)?  If so, it would be helpful to have a comment here indicating that validation of the index must be done by the caller.  And if that's the case then why have this check (and short-circuit) here?

File: Modules/pyexpat.c [hunk 4]
```diff
@@ -1682,8 +1690,8 @@ MODULE_INITFUNC(void)
         Py_DECREF(m);
         return NULL;
     }
-    errors_module = PyDict_GetItem(d, errmod_name);
-    if (errors_module == NULL) {
+    errors_module = PyDict_GetItemWithError(d, errmod_name);
+    if (errors_module == NULL && !PyErr_Occurred()) {
```
Comment (ericsnowcurrently) [inline Modules/pyexpat.c:1694]: Doesn't the error need to be returned or cleared?

File: Modules/pyexpat.c [hunk 5]
```diff
@@ -1692,7 +1700,7 @@ MODULE_INITFUNC(void)
         }
     }
     Py_DECREF(errmod_name);
-    model_module = PyDict_GetItem(d, modelmod_name);
+    model_module = PyDict_GetItemWithError(d, modelmod_name);
```
Comment (ericsnowcurrently) [inline Modules/pyexpat.c:1703]: Doesn't the error need to be returned or cleared?

File: Objects/dictobject.c (hunk 1 above)
Comment (ericsnowcurrently) [inline Objects/dictobject.c]: This is a little hard to read.  There's a lot going on in these two lines.  Perhaps split it up a little?

File: Objects/moduleobject.c [hunk 6]
```diff
@@ -792,16 +792,17 @@ static PyObject *
...
 
     if (dict != NULL) {
         if (PyDict_Check(dict)) {
-            PyObject *dirfunc = PyDict_GetItemString(dict, "__dir__");
+            PyObject *dirfunc = _PyDict_GetItemIdWithError(dict, &PyId___dir__);
             if (dirfunc) {
                 result = _PyObject_CallNoArg(dirfunc);
             }
-            else {
+            else if (!PyErr_Occurred()) {
```
Comment (ericsnowcurrently) [inline Objects/moduleobject.c:805]: How about clearing or raising the exception if there is one?

File: Objects/object.c [hunk 7]
```diff
@@ -17,7 +17,6 @@ _Py_IDENTIFIER(Py_Repr);
 _Py_IDENTIFIER(__bytes__);
 _Py_IDENTIFIER(__dir__);
 _Py_IDENTIFIER(__isabstractmethod__);
-_Py_IDENTIFIER(builtins);
```
Comment (ericsnowcurrently) [inline Objects/object.c]: Is this just a drive-by change?  Otherwise I don't see how it fits.

File: Objects/object.c [hunk 8]
```diff
@@ -1250,13 +1255,23 @@ _PyObject_GenericGetAttrWithDict(PyObject *obj, PyObject *name,
...
         if (res != NULL) {
             Py_INCREF(res);
             Py_DECREF(dict);
             goto done;
         }
-        Py_DECREF(dict);
+        else {
+            Py_DECREF(dict);
+            if (PyErr_Occurred()) {
+                if (suppress && PyErr_ExceptionMatches(PyExc_AttributeError)) {
```
Comment (ericsnowcurrently) [inline Objects/object.c:1267]: good catch

File: Objects/object.c [hunk 9]
```diff
@@ -2094,7 +2112,7 @@ Py_ReprLeave(PyObject *obj)
     if (dict == NULL)
         goto finally;
 
-    list = _PyDict_GetItemId(dict, &PyId_Py_Repr);
+    list = _PyDict_GetItemIdWithError(dict, &PyId_Py_Repr);
```
Comment (ericsnowcurrently) [inline Objects/object.c:1998]: This change is an improvement, but note that it is mostly unnecessary since the exception is cleared below in `finally`.

File: Objects/typeobject.c [hunk 10]
```diff
@@ -2807,8 +2841,12 @@ type_new(PyTypeObject *metatype, PyObject *args, PyObject *kwds)
             goto error;
         }
         PyCell_Set(cell, (PyObject *) type);
-        _PyDict_DelItemId(dict, &PyId___classcell__);
-        PyErr_Clear();
+        if (_PyDict_DelItemId(dict, &PyId___classcell__) < 0) {
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:2844]: nice

File: Objects/typeobject.c [hunk 11]
```diff
@@ -3709,47 +3747,38 @@ object_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
...
-        sorted_methods = PyObject_CallFunctionObjArgs(sorted,
-                                                      abstract_methods,
-                                                      NULL);
+            return NULL;
+        sorted_methods = PySequence_List(abstract_methods);
+        Py_DECREF(abstract_methods);
         if (sorted_methods == NULL)
-            goto error;
+            return NULL;
+        if (PyList_Sort(sorted_methods)) {
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:3765]: Is this strictly equivalent to calling the `sorted` function looked up on the `builtins` module?  I would expect so.

Also, presumably there isn't a `PyObject_Sort()` or `PySequence_Sort()` we could use to avoid the cost of copying into a new list.  On the other hand, using a copy makes a lot more sense. :)

File: Objects/typeobject.c [hunk 12]
```diff
@@ -3709,47 +3747,38 @@ object_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
...
+            Py_DECREF(sorted_methods);
+            return NULL;
+        }
         comma = _PyUnicode_FromId(&comma_id);
         if (comma == NULL)
-            goto error;
+            return NULL;
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:3772]: Doesn't `sorted_methods` need to be decref'ed before returning?

File: Objects/typeobject.c [hunk 13]
```diff
@@ -3709,47 +3747,38 @@ object_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
...
         joined = PyUnicode_Join(comma, sorted_methods);
         if (joined == NULL)
-            goto error;
+            return NULL;
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:3777]: Doesn't `sorted_methods` need to be decref'ed before returning?  I suppose you could do it right after calling `PyUnicode_Join()`.

File: Objects/typeobject.c [hunk 14]
```diff
@@ -4578,14 +4607,12 @@ object___reduce_ex___impl(PyObject *self, int protocol)
     if (objreduce == NULL) {
         objreduce = _PyDict_GetItemId(PyBaseObject_Type.tp_dict,
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:4611]: What about `_PyDict_GetItemIdWithError()`?

File: Objects/typeobject.c [hunk 15]
```diff
@@ -4578,14 +4607,12 @@ object___reduce_ex___impl(PyObject *self, int protocol)
...
                                       &PyId___reduce__);
-        if (objreduce == NULL)
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:4581]: Did you mean to remove this?  I don't see an equivalent replacement in the PR.

File: Objects/typeobject.c [hunk 16]
```diff
@@ -4578,14 +4607,12 @@ object___reduce_ex___impl(PyObject *self, int protocol)
...
-            return NULL;
     }
 
-    reduce = _PyObject_GetAttrId(self, &PyId___reduce__);
-    if (reduce == NULL)
-        PyErr_Clear();
-    else {
+    if (_PyObject_LookupAttrId(self, &PyId___reduce__, &reduce) < 0) {
+        return NULL;
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:4616]: This isn't equivalent, is it?  In the original code the error is cleared and it falls back to calling `_common_reduce()`.

File: Objects/typeobject.c [hunk 17]
```diff
@@ -4797,14 +4824,12 @@ static int
 add_methods(PyTypeObject *type, PyMethodDef *meth)
 {
     PyObject *dict = type->tp_dict;
+    PyObject *name;
 
     for (; meth->ml_name != NULL; meth++) {
         PyObject *descr;
         int err;
         int isdescr = 1;
-        if (PyDict_GetItemString(dict, meth->ml_name) &&
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:4805]: Why move this down in the function?  It seems like that might cause a slight change in semantics.

File: Objects/typeobject.c [hunk 18]
```diff
@@ -4797,14 +4824,12 @@ static int
 add_methods(PyTypeObject *type, PyMethodDef *meth)
 {
     PyObject *dict = type->tp_dict;
+    PyObject *name;
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:4830]: I'd put this down where it's first used (i.e. right above line 4852 `if (isdescr) {`).

File: Objects/typeobject.c [hunk 19]
```diff
@@ -4826,11 +4851,36 @@ add_methods(PyTypeObject *type, PyMethodDef *meth)
...
+            if (name == NULL) {
+                Py_DECREF(descr);
+                return -1;
+            }
+        }
+
+        if (!(meth->ml_flags & METH_COEXIST)) {
+            if (PyDict_GetItemWithError(dict, name)) {
+                if (!isdescr) {
+                    Py_DECREF(name);
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:4872]: ...or use `Py_XDECREF`.

File: Objects/typeobject.c [hunk 20]
```diff
@@ -4845,12 +4895,18 @@ add_members(PyTypeObject *type, PyMemberDef *memb)
     PyObject *dict = type->tp_dict;
 
     for (; memb->name != NULL; memb++) {
-        PyObject *descr;
-        if (PyDict_GetItemString(dict, memb->name))
-            continue;
-        descr = PyDescr_NewMember(type, memb);
+        PyObject *descr = PyDescr_NewMember(type, memb);
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:4901]: It seems a little strange to break the short-circuit like this (here and in a few other places in this PR), but I see your intent: use `GetItem` instead of `GetItemString`.  As long as the performance (and semantics) impact is minimal that should be fine.

File: Python/ceval.c [hunk 21]
```diff
@@ -5109,10 +5138,13 @@ unicode_concatenate(PyObject *v, PyObject *w,
             PyObject *names = f->f_code->co_names;
             PyObject *name = GETITEM(names, oparg);
             PyObject *locals = f->f_locals;
-            if (locals && PyDict_CheckExact(locals) &&
-                PyDict_GetItem(locals, name) == v) {
-                if (PyDict_DelItem(locals, name) != 0) {
-                    PyErr_Clear();
+            if (locals && PyDict_CheckExact(locals)) {
+                PyObject *w = PyDict_GetItemWithError(locals, name);
+                if ((w == v && PyDict_DelItem(locals, name) != 0) ||
```
Comment (ericsnowcurrently) [inline Python/ceval.c:5198]: For what it's worth, I find this more readable than the similar code I complained about above. :)

File: Python/getargs.c [hunk 22]
```diff
@@ -1990,13 +1997,10 @@ parser_clear(struct _PyArg_Parser *parser)
 }
 
 static PyObject*
-find_keyword(PyObject *kwargs, PyObject *kwnames, PyObject *const *kwstack, PyObject *key)
+find_keyword(PyObject *kwnames, PyObject *const *kwstack, PyObject *key)
```
Comment (ericsnowcurrently) [inline Python/getargs.c:2024]: +1 on simplifying the signature :)

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
First of all, thanks for working on this! :)  Overall it looks good.

My main concern with this PR is changing semantics.  From what I can tell you're introducing a bunch of changes in behavior, albeit it corner error cases.  What is the risk to compatibility?  My gut tells me there's at least a slight risk.

Secondly, you've touched a lot of critical code.  Please make sure to run the benchmark suite to ensure the PR doesn't slow down Python. :)

Also, there a number of places where I wanted to suggest a better spelling.  However, such changes would be slightly riskier and would mostly clutter up the PR, obscuring the core changes.  So I've left out those comments and focused mostly on checking correctness.

Finally, the most likely thing I might have missed in this review is refcounts.  You've added quite a few places that exit early when there's an error.  I'm not sure that I checked to make sure everything was properly decref'ed in those new error cases.

**Comment by ericsnowcurrently** [comment]
Incidentally, how many uses of the non-`WithError` API remain after this?

File: Modules/_sre.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Modules/_sre.c]: The removed code was moved into `match_getindex()` because it was repeated after every call of `match_getindex()`. This function is only called with valid index or the result of `match_getindex()`. This check was here just to make the caller place simpler. Will move it to the caller place and add an assert instead.

File: Modules/_csv.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Modules/_csv.c:1437]: Yes, since `PyDict_DelItem()` is used often with `PyDict_GetItem()` I looked also at its usage. The general principle is the same -- replace only expected exceptions, and keep other exceptions unchanged. In this case replacing KeyError with csv.error is intentional, other exceptions signal about errors.

File: Modules/pyexpat.c (hunk 4 above)
Comment (serhiy-storchaka) [reply Modules/pyexpat.c:1694]: It is returned below, at line 1713.

File: Modules/pyexpat.c (hunk 5 above)
Comment (serhiy-storchaka) [reply Modules/pyexpat.c:1703]: It is returned below, at line 1713.

File: Objects/dictobject.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Objects/dictobject.c]: Good catch!

Comment (serhiy-storchaka) [reply Objects/dictobject.c]: Done.

File: Objects/moduleobject.c (hunk 6 above)
Comment (serhiy-storchaka) [reply Objects/moduleobject.c:805]: It is returned below, at line 819.

File: Objects/object.c (hunk 7 above)
Comment (serhiy-storchaka) [reply Objects/object.c]: It is a fix of the compiler warning introduced in previous commit. It is already fixed in a separate PR, so this change will gone after merging with master.

File: Objects/object.c (hunk 9 above)
Comment (serhiy-storchaka) [reply Objects/object.c:1998]: Yes, I made changes even in such cases, because I want to decrease the number of suspicious `PyDict_GetItem()`, and because `PyDict_GetItemWithError()` is faster.

File: Objects/typeobject.c (hunk 11 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c:3765]: Yes, this is strictly equivalent. Just removed an overhead for lookup and function call.

Other occurrences of `PyEval_GetBuiltins()` + `PyDict_GetItem()` were fixed in other issue (there was a bug, because the result was not checked), and I left this one for this PR.

The `sorted` builtin makes a copy too. The pair `PySequence_List()` + `PyList_Sort()` is used also in other places.

File: Objects/typeobject.c (hunk 12 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c:3772]: Good catch!

File: Objects/typeobject.c (hunk 13 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c:3777]: Good catch!

File: Objects/typeobject.c (hunk 15 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c:4581]: This code was definitely incorrect. It returned NULL without setting an exception. But this should not happen on practice. And in worst and unlikely case, we will just not use a fast path in the code below if objreduce == NULL.

File: Objects/typeobject.c (hunk 14 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c:4611]: This call should be always successful. It is easier to keep `_PyDict_GetItemId()` and ignore an error than add the handling of errors here. Maybe this will be changed in other attempt.

File: Objects/typeobject.c (hunk 16 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c:4616]: It will fall back in the case of AttributeError. All other errors are unexpected and should be propagated.

File: Objects/typeobject.c (hunk 17 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c:4805]: There were only two cases in the stdlib (besides tests) with duplicated names. And one of them was already fixed in separate issue. In all other cases it is more efficient to check for duplicates after creation of the descriptor object (we can reuse a string object).

File: Objects/typeobject.c (hunk 19 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c:4872]: No, `name` is a borrowed reference if `isdescr` is true.

File: Objects/typeobject.c (hunk 20 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c:4901]: Actually this improves the performance by avoiding to create a temporary string object. The "short-circuit" is occurred in very very rare cases.

File: Python/getargs.c (hunk 22 above)
Comment (serhiy-storchaka) [reply Python/getargs.c:2024]: This is because the part of this function was moved to the caller place. This complicates the code, but avoids to call redundant `PyErr_Occurred()` in performance critical code.

**Review by serhiy-storchaka — COMMENTED** [review]
Thank you for your review @ericsnowcurrently!

File: Modules/pyexpat.c (hunk 5 above)
Comment (ericsnowcurrently) [reply Modules/pyexpat.c:1703]: The model_module gets reset on line 1705, so the error from the first attempt may remain uncleared or swallowed

Comment (serhiy-storchaka) [reply Modules/pyexpat.c:1703]: Line 1705 is executed only no error was raised at this line.

Error handling in this function (as well as in many other module initialization functions) is pretty poor. Results of `PyModule_AddObject()` and derived functions are not checked, and references are leaked in case of error. But this is different issue(s).

Comment (ericsnowcurrently) [reply Modules/pyexpat.c:1703]: Then shouldn't line 1704 have ` && !PyErr_Occurred()` like line 1694 does?

File: Modules/pyexpat.c (hunk 4 above)
Comment (ericsnowcurrently) [reply Modules/pyexpat.c:1694]: got it

File: Objects/moduleobject.c (hunk 6 above)
Comment (ericsnowcurrently) [reply Objects/moduleobject.c:805]: sorry, I missed that

File: Objects/typeobject.c (hunk 16 above)
Comment (ericsnowcurrently) [reply Objects/typeobject.c:4616]: ah, that makes sense

File: Modules/pyexpat.c (hunk 5 above)
Comment (serhiy-storchaka) [reply Modules/pyexpat.c:1703]: Already added.

**Comment by serhiy-storchaka** [comment]
As for benchmarks, running the benchmark suite exposes some slowdown on some tests, but results can have significant random component. I'll research this in more details to get more trustworthy result.

**Comment by serhiy-storchaka** [comment]
I have made the requested changes; please review again.
